# **Maestría en Inteligencia Artificial Aplicada**

## **Curso: Inteligencia Artificial y Aprendizaje Automático**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

## **Actividad de las Semanas 5 y 6**
### **Problema de asignación de créditos: South German Dataset.**


## **Nombres y matrículas:**


🔹 Jossie Paola Jiménez Rivera   ➡️ A01795961

🔹 Javier Augusto Rebull Saucedo ➡️ A01795838

🔹 Nancy Teresa Zapién García    ➡️ A01795907



# **Parte I: Partición, análisis y pre-procesamiento de los datos.**

## **Ejercicio 1**

In [3]:
# Aquí deberás incluir todas las librerías que requieras durante esta actividad:

#Para cargar archivos
from google.colab import drive
import os

# Para manipular datos
import numpy as np  # Biblioteca para realizar operaciones numéricas y trabajar con arreglos multidimensionales.
import pandas as pd  # Biblioteca para manipulación y análisis de datos, ofrece estructuras como DataFrame y Series.

# Ejercicio 3.
from sklearn.model_selection import train_test_split  # Función para dividir los datos en conjuntos de entrenamiento y prueba.

# Ejercicio 5.
from sklearn.pipeline import Pipeline  # Clase que ayuda a ensamblar varios pasos de procesamiento y modelado en una secuencia.
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, OrdinalEncoder  # Herramientas para la transformación de características: escalado y codificación de variables categóricas.
from sklearn.compose import ColumnTransformer  # Permite aplicar diferentes transformaciones a diferentes columnas en un conjunto de datos.
from sklearn.impute import SimpleImputer  # Rellena valores faltantes en los datos según una estrategia específica (como la media, mediana, moda).

# Ejercicio 7.
from sklearn.linear_model import LogisticRegression  # Modelo de regresión logística para problemas de clasificación binaria y multiclase.
from sklearn.tree import DecisionTreeClassifier  # Modelo de clasificación basado en árboles de decisión.
from sklearn.neighbors import KNeighborsClassifier  # Algoritmo de clasificación basado en vecinos más cercanos.
from sklearn.ensemble import RandomForestClassifier  # Modelo de ensamble basado en múltiples árboles de decisión para mejorar la precisión.
from xgboost import XGBClassifier  # Modelo de clasificación de gradiente boosting basado en árboles, conocido por su alto rendimiento.
from sklearn.neural_network import MLPClassifier  # Modelo de red neuronal multicapa para clasificación.
from sklearn.svm import SVC  # Máquina de vectores de soporte para clasificación.

# Para conjuntos desbalanceados
from imblearn.over_sampling import SMOTE, KMeansSMOTE  # Técnicas de sobremuestreo para equilibrar clases desbalanceadas en problemas de clasificación.
from sklearn.model_selection import RepeatedStratifiedKFold, cross_validate  # Métodos para validación cruzada y división estratificada repetida en conjuntos de datos.
from sklearn.metrics import make_scorer  # Función que permite crear métricas personalizadas para la evaluación de modelos.
from imblearn.metrics import geometric_mean_score  # Métrica de rendimiento para conjuntos de datos desbalanceados, basada en la media geométrica.

In [2]:
# Si se desean comentar algunos de los Warnings.
import warnings  # Módulo para controlar los warnings que surgen durante la ejecución del código.
warnings.filterwarnings('ignore')  # Esta línea desactiva la visualización de warnings en el código. Se usa para evitar que se impriman mensajes de advertencia que no sean críticos y así tener una salida

In [4]:
# Carga y renombra los nombres de las columnas del alemán al inglés y desplegamos
# de nuevo el DataFrame para ver el resultado obtenido:

# ************* Incluye aquí tu código:*****************************
# NOTA: Tanto aquí como en lo sucesivo, "None" puede indicar una o varias líneas de código.

drive.mount('/content/drive')  # Monta Google Drive para acceder a archivos desde Colab.

DIR = '/content/drive/MyDrive/TecMTY/septiembre noviembre 2024/IA y ML/Semana 5 y 6'
os.chdir(DIR)  # Cambia el directorio a la ubicación del archivo en Google Drive.

# Lectura del archivo .asc
df = pd.read_csv('SouthGermanCredit.asc', sep=' ')  # Lee el archivo de crédito alemán con separación por espacio.

# Renombra columnas de Alemán a Inglés
columns = [
    'status',                  # Status (laufkont)
    'duration',                # Duration (laufzeit)
    'credit_history',          # Credit history (moral)
    'purpose',                 # Purpose (verw)
    'amount',                  # Amount (hoehe)
    'savings',                 # Savings (sparkont)
    'employment_duration',     # Employment duration (beszeit)
    'installment_rate',        # Installment rate (rate)
    'personal_status_sex',     # Personal status sex (famges)
    'other_debtors',           # Other debtors (buerge)
    'present_residence',       # Present residence (wohnzeit)
    'property',                # Property (verm)
    'age',                     # Age (alter)
    'other_installment_plans', # Other installment plans (weitkred)
    'housing',                 # Housing (wohn)
    'number_credits',          # Number credits (bishkred)
    'job',                     # Job (beruf)
    'people_liable',           # People liable (pers)
    'telephone',               # Telephone (telef)
    'foreign_worker',          # Foreign worker (gastarb)
    'credit_risk'              # Credit risk (kredit)
]

# Guarda nuevos nombres de las columnas al DataFrame
df.columns = columns  # Asigna la lista 'columns' como nombres de las columnas del DataFrame.

# *********** Aquí termina la sección de agregar código *************

df.head().T  # Muestra las primeras filas del DataFrame transpuestas para visualizar mejor los nombres de las columnas y valores.

Mounted at /content/drive


0     1    2     3     4
status                      1     1    2     1     1
duration                   18     9   12    12    12
credit_history              4     4    2     4     4
purpose                     2     0    9     0     0
amount                   1049  2799  841  2122  2171
savings                     1     1    2     1     1
employment_duration         2     3    4     3     3
installment_rate            4     2    2     3     4
personal_status_sex         2     3    2     3     3
other_debtors               1     1    1     1     1
present_residence           4     2    4     2     4
property                    2     1    1     1     2
age                        21    36   23    39    38
other_installment_plans     3     3    3     3     1
housing                     1     1    1     1     2
number_credits              1     2    1     2     2
job                         3     3    2     2     2
people_liable               2     1    2     1     2
telephone                   1     1    1     1     1
foreign_worker              2     2    2     1     1
credit_risk                 1     1    1     1     1

## **Ejercicio 2**

In [5]:
# Realiza a continuación una transformación para que la clase negativa (buen cliente)
# quede con el valor de 0 y la clase positiva (mal cliente) quede con el valor de 1.

# ************* Incluye aquí tu código:*****************************
# Intercambia los valores de la columna 'credit_risk': los buenos clientes (0) pasan a ser 1, y los malos clientes (1) pasan a ser 0.
df['credit_risk'] = df['credit_risk'].replace({0:1, 1:0})

# *********** Aquí termina la sección de agregar código *************

print(df['credit_risk'].value_counts())  # Muestra el conteo de valores de la columna 'credit_risk' después de la transformación para verificar el cambio.

credit_risk
0    700
1    300
Name: count, dtype: int64


## **Ejercicio 3**

In [6]:
# Realiza una partición solicitada de entrenamiento y prueba.
# Los nombres de los conjuntos deberán ser como se indican en los print de abajo:

# ************* Incluye aquí tu código:*****************************

X = df.drop(columns='credit_risk')  # Características (variables independientes) del conjunto de datos, excluyendo la variable de salida 'credit_risk'.
y = df['credit_risk']  # Variable de salida (variable dependiente), en este caso, 'credit_risk'.

# El artículo sugiere una partición de (70:30), Entrenamiento 70% y Prueba 30%
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, stratify=y, random_state=1)
# Se realiza la división en conjuntos de entrenamiento y prueba, manteniendo la proporción de las clases (stratify) y con un random_state para reproducibilidad.

# *********** Aquí termina la sección de agregar código *************

# Mostremos las dimensiones de la partición generada:
print("Dimensiones:")  # Se muestran las dimensiones de los conjuntos de entrenamiento y prueba.
print("Entrenamiento:", Xtrain.shape, ytrain.shape)  # Dimensiones del conjunto de entrenamiento.
print("Prueba:", Xtest.shape, ytest.shape)  # Dimensiones del conjunto de prueba.

# Y el porcentaje de cada clase de la variable de salida:
tmp = ytrain.sum() / ytrain.shape[0]  # Cálculo del porcentaje de la clase positiva en el conjunto de entrenamiento.
print("\nPorcentaje clases Positiva:%.2f%%, y Negativa:%.2f%%" % (100*(1-tmp), tmp*100))  # Se imprime el porcentaje de clase positiva y negativa.

Dimensiones:
Entrenamiento: (700, 20) (700,)
Prueba: (300, 20) (300,)

Porcentaje clases Positiva:70.00%, y Negativa:30.00%


### **Con base al porcentaje de los niveles de la variable de salida ¿podemos decir que tenemos un problema de datos desbalanceado? ¿Por qué?**

++++++++ Inicia la sección de agregar texto: +++++++++++

Con base en los porcentajes obtenidos, podemos confirmar que existe un desbalance en el conjunto de datos. Del 100% de los datos, el 70% corresponde a la clase positiva (**malos clientes**) y el 30% a la clase negativa (**buenos clientes**). Este desbalance puede impactar negativamente el rendimiento de los modelos de aprendizaje automático, ya que tienden a ser más efectivos en identificar la clase mayoritaria (malos clientes), mientras que su precisión al predecir la clase minoritaria (buenos clientes) disminuye.

Cuando se trabaja con datos desbalanceados, es común que el modelo se ajuste para predecir mejor la clase con mayor representación (0), lo que puede llevar a errores significativos al identificar casos de la clase minoritaria (1). Esto representa un reto importante para la construcción de modelos que busquen un buen equilibrio entre ambas clases.

Para mitigar este problema, es crucial aplicar técnicas adecuadas como el submuestreo de la clase mayoritaria, el sobremuestreo de la clase minoritaria, o el uso de algoritmos diseñados para manejar conjuntos de datos desbalanceados. Con estas estrategias, podemos mejorar la capacidad del modelo para identificar correctamente los clientes de la clase minoritaria sin comprometer su rendimiento general.

### ++++++++ Termina la sección de agregar texto: +++++++++++

## **Ejercicio 4**

In [7]:
# De acuerdo a la información de la Tabla 3 del artículo de la IEEE
# define las variables correspondientes que se indican a continuación:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
lista_paper_num = [
    'duration',        # Duración del crédito
    'amount',          # Monto total del crédito
    'age',             # Edad del cliente
    'peope_liable'     # Número de personas a cargo del cliente
]
# 4 variables numéricas

# Variables ordinales:
lista_paper_ord = [
    'employment_duration',    # Duración del empleo del cliente
    'installment_rate',       # Tasa de las cuotas del crédito
    'present_residence',      # Tiempo de residencia en el domicilio actual
    'property',               # Tipo de propiedad
    'number_credits',         # Número de créditos anteriores
    'job'                     # Ocupación del cliente
]
# 6 variables ordinales

# Variables nominales:
lista_paper_cat = [
    'status',                 # Estado del crédito
    'credit_history',         # Historial crediticio
    'purpose',                # Propósito del crédito
    'savings',                # Ahorros del cliente
    'personal_status_sex',    # Estado civil y sexo
    'other_debtors',          # Otros deudores o avales
    'other_installment_plans',# Otros planes de cuotas
    'housing',                # Tipo de vivienda
    'telephone',              # Disponibilidad de teléfono
    'foreign_worker'          # Si el cliente es trabajador extranjero
]
# 10 variables nominales (no se toma en cuenta 'credit_risk')

# *********** Aquí termina la sección de agregar código *************

## **Ejercicio 5**

In [8]:
# Transformaciones que se aplicarán a las variables numéricas usando la clase Pipeline de sklearn:

# ************* Incluye aquí tu código:*****************************

# Variables numéricas:
numericas_pipe = Pipeline([
    ("impute", SimpleImputer(strategy="median")),   # Imputar (rellenar) valores faltantes con la mediana de cada variable numérica.
    ("standardize", MinMaxScaler())                 # Escalar los valores numéricos para que estén en un rango entre 0 y 1, usando MinMaxScaler.
])
numericas_pipe_nombres = ['duration', 'amount', 'age', 'people_liable']  # Lista de los nombres de las variables numéricas a transformar.

# Variables categóricas-Nominales:
nominales_pipe = Pipeline([
    ("impute", SimpleImputer(strategy="most_frequent")),  # Imputar valores faltantes con el valor más frecuente (moda) para variables nominales.
    ("encoder", OneHotEncoder(handle_unknown="ignore", drop="first"))  # Codificación One-Hot, ignora categorías desconocidas y evita la colinealidad (drop="first").
])
nominales_pipe_nombres = ['status', 'credit_history', 'purpose', 'savings',
                          'personal_status_sex', 'other_debtors',
                          'other_installment_plans', 'housing', 'telephone', 'foreign_worker']
# Lista de los nombres de las variables categóricas nominales a transformar (excluye 'credit_risk').

# Variables categóricas-ordinales:
ordinales_pipe = Pipeline([
    ("impute", SimpleImputer(strategy="most_frequent")),  # Imputar valores faltantes con el valor más frecuente.
    ("encoder", OrdinalEncoder())                        # Codificación ordinal, que asigna valores numéricos ordenados a las categorías.
])
ordinales_pipe_nombres = ['employment_duration', 'installment_rate', 'present_residence',
                          'property', 'number_credits', 'job']
# Lista de los nombres de las variables categóricas ordinales a transformar.

# Conjunta las transformaciones de todo tipo de variable y
# deja sin procesar aquellas que hayas decidido no transformar:
columnasTransformer = ColumnTransformer(transformers=[
    ('num', numericas_pipe, numericas_pipe_nombres),  # Aplica las transformaciones numéricas a las variables numéricas.
    ('ord', ordinales_pipe, ordinales_pipe_nombres),  # Aplica las transformaciones ordinales a las variables categóricas ordinales.
    ('nom', nominales_pipe, nominales_pipe_nombres)   # Aplica las transformaciones nominales a las variables categóricas nominales.
], remainder='passthrough')  # Mantiene las columnas no especificadas sin procesar ('passthrough').

# Explicación:
# - Pipelines: Se utilizan para aplicar secuencias de transformaciones sobre conjuntos de variables. Por ejemplo,
#   las variables numéricas primero rellenan valores faltantes con la mediana y luego se escalan.
# - ColumnTransformer: Permite aplicar diferentes transformaciones a diferentes subconjuntos de variables
#   (numéricas, ordinales y nominales) al mismo tiempo.
# - 'remainder="passthrough"': Indica que las columnas no especificadas en el ColumnTransformer no se transforman
#   y se mantienen tal como están.

# *********** Aquí termina la sección de agregar código *************

## **Ejercicio 6**

In [9]:
# Como se va a utilizar Validación-Cruzada, concatena los conjuntos de entrenamiento y prueba
# en uno nuevo conjunto aumentado que llamaremos trainval para utilizar como entrenamiento:

# ************* Incluye aquí tu código:**************************

# Se concatenan los conjuntos de entrenamiento (Xtrain, ytrain) y prueba (Xtest, ytest) para formar un único conjunto
# combinado llamado Xtraintest y ytraintest. Esto es útil cuando se desea entrenar y validar un modelo con
# todo el conjunto de datos disponible, aplicando Validación Cruzada.
Xtraintest = pd.concat([Xtrain, Xtest], axis=0)  # Concatena las características (features) de los conjuntos de entrenamiento y prueba.
ytraintest = pd.concat([ytrain, ytest], axis=0)  # Concatena las etiquetas (target) de los conjuntos de entrenamiento y prueba.

# *********** Aquí termina la sección de agregar código *************

# Veamos cuántas variables nuevas se introducen con las transformaciones One-Hot-Encoding:
Xtmp = Xtraintest.copy()  # Se realiza una copia de Xtraintest para aplicar las transformaciones y no modificar los datos originales.
tmp = columnasTransformer.fit_transform(Xtmp)  # Se ajusta y transforma el conjunto de datos usando el ColumnTransformer, que aplica One-Hot-Encoding y otras transformaciones.
print("Dimensión de las variables de entrada ANTES de las transformaciones:", Xtmp.shape)  # Imprime la dimensión original del conjunto de datos antes de las transformaciones.
print("Dimensión de las variables de entrada DESPUÉS de las transformaciones:", tmp.shape)  # Imprime la dimensión del conjunto de datos después de aplicar las transformaciones.

Dimensión de las variables de entrada ANTES de las transformaciones: (1000, 20)
Dimensión de las variables de entrada DESPUÉS de las transformaciones: (1000, 41)


## **Ejercicio 7**

In [10]:
# Definimos a continuación la función que llamamos "mis_modelos" que incluye
# todos los modelos que deseamos comparar en el ejercicio.

def mis_modelos():
    # Inicializamos listas para almacenar los modelos y sus nombres
    modelos, nombres = list(), list()

    # ************* Incluye aquí tu código:**************************
    #
    # Deberás incluir en cada modelo los argumentos que consideres
    # adecuados para que cada uno converja y no esté sobreentrenado
    # con respecto a la métrica de la exactitud (accuracy).

    # Regresión Logística - Logistic Regression (LR):
    modelos.append(LogisticRegression(penalty='elasticnet',
                                      solver='saga',
                                      tol=0.0001,         # Tolerancia
                                      fit_intercept=True,
                                      max_iter=2000,
                                      C=10,
                                      l1_ratio=0.34,
                                      random_state=1))
    nombres.append('LR')

    # k-Vecinos más Cercanos - k-Nearest-Neighbors (kNN):
    modelos.append(KNeighborsClassifier(n_neighbors=20,
                                        p=1,
                                        leaf_size=35,
                                        n_jobs=-1,
                                        weights="uniform"))
    nombres.append('kNN')

    # Árbol de Decisiones - DecisionTree (DTree):
    modelos.append(DecisionTreeClassifier(criterion="entropy",
                                          max_depth=10,
                                          min_samples_split=15,
                                          random_state=1))
    nombres.append('DTree')

    # Bosque Aleatorio - RandomForest (RF):
    modelos.append(RandomForestClassifier(n_estimators=200,
                                          criterion="entropy",
                                          max_features="sqrt",
                                          random_state=1,
                                          max_depth=10))
    nombres.append('RF')

    # XGBoost:
    modelos.append(XGBClassifier(random_state=1,
                                 tree_method="hist",
                                 learning_rate=0.05,
                                 max_depth=7))
    nombres.append('XGBoost')

    # Red Neuronal de Perceptrón Multicapa - MLP:
    modelos.append(MLPClassifier(solver="adam",
                                 random_state=1,
                                 hidden_layer_sizes=32,
                                 activation="identity",
                                 max_iter=2500,
                                 alpha=0.001))
    nombres.append('MLP')

    # Máquina de Vectores de Soporte - SVM:
    modelos.append(SVC(C=14,
                       kernel="linear",
                       tol=0.0001))
    nombres.append('SVM')

    return modelos, nombres

# Técnica de submuestreo (undersampling) y/o sobremuestreo (oversampling) utilizada:
mi_uoSampling = KMeansSMOTE(random_state=1, k_neighbors=6, n_jobs=-1)

# Aplicamos el sobremuestreo a los datos de entrenamiento:
Xtv_uo, ytv_uo = mi_uoSampling.fit_resample(Xtraintest, ytraintest)

# *********** Aquí termina la sección de agregar código *******************

# Entrenamos cada uno de los modelos y desplegamos las métricas de Train y Validation (Val).
# NOTA: Observa que el método de Validación-Cruzada llama a los resultados de "validation" como "test".

modelos, nombres = mis_modelos()  # Llamamos a la función para obtener los modelos y nombres
resultados = list()

# Iteramos sobre cada modelo para entrenarlo y evaluar su rendimiento.
for i in range(len(modelos)):
    # Definimos el pipeline con las transformaciones y los modelos
    pipeline = Pipeline(steps=[('ct', columnasTransformer), ('m', modelos[i])])

    # Aplicamos validación-cruzada con 5 divisiones repetidas 3 veces
    micv = RepeatedStratifiedKFold(n_splits=5, n_repeats=3, random_state=5)

    # Definimos las métricas que deseamos recuperar
    mismetricas = {
        'miaccuracy': 'accuracy',
        'miprecision': 'precision',
        'mirecall': 'recall',
        'mifi': 'f1',
        'miauc': 'roc_auc',
        'migmean': make_scorer(geometric_mean_score)
    }

    # Entrenamos el modelo usando validación cruzada
    scores = cross_validate(pipeline, Xtv_uo, ytv_uo, scoring=mismetricas, cv=micv, return_train_score=True)

    # Guardamos los resultados del modelo para análisis posteriores
    resultados.append(scores)

    # Desplegamos las métricas para verificar posibles problemas de subentrenamiento o sobreentrenamiento
    print('>> %s' % nombres[i])
    for j, k in enumerate(list(scores.keys())):
        if j > 1:
            print('\t %s %.3f (%.3f)' % (k, np.mean(scores[k]), np.std(scores[k])))

# El código entrena cada uno de los modelos, aplica validación cruzada y muestra las métricas
# para evaluar el rendimiento del modelo en los datos de entrenamiento y validación.

>> LR
	 test_miaccuracy 0.817 (0.029)
	 train_miaccuracy 0.832 (0.007)
	 test_miprecision 0.827 (0.034)
	 train_miprecision 0.845 (0.010)
	 test_mirecall 0.803 (0.036)
	 train_mirecall 0.813 (0.007)
	 test_mifi 0.814 (0.030)
	 train_mifi 0.829 (0.007)
	 test_miauc 0.903 (0.021)
	 train_miauc 0.921 (0.005)
	 test_migmean 0.816 (0.030)
	 train_migmean 0.832 (0.007)
>> kNN
	 test_miaccuracy 0.795 (0.020)
	 train_miaccuracy 0.817 (0.006)
	 test_miprecision 0.840 (0.022)
	 train_miprecision 0.865 (0.011)
	 test_mirecall 0.731 (0.032)
	 train_mirecall 0.751 (0.008)
	 test_mifi 0.781 (0.023)
	 train_mifi 0.804 (0.006)
	 test_miauc 0.881 (0.022)
	 train_miauc 0.907 (0.005)
	 test_migmean 0.793 (0.021)
	 train_migmean 0.814 (0.006)
>> DTree
	 test_miaccuracy 0.792 (0.026)
	 train_miaccuracy 0.888 (0.009)
	 test_miprecision 0.802 (0.034)
	 train_miprecision 0.910 (0.021)
	 test_mirecall 0.777 (0.039)
	 train_mirecall 0.862 (0.030)
	 test_mifi 0.789 (0.027)
	 train_mifi 0.885 (0.010)
	 test_miauc

## **Ejercicio 8**

### **Escribe tus conclusiones finales de la actividad.**

++++++++ Inicia la sección de agregar texto: +++++++++++

### **Conclusiones Finales de la Actividad:**

En esta actividad se evaluaron diferentes modelos de aprendizaje automático y se implementaron técnicas de sobre-muestreo (SMOTE) para abordar el problema de desbalance de clases. El desbalance de clases, donde la clase mayoritaria (buenos clientes) representa el 70% y la clase minoritaria (malos clientes) el 30%, puede afectar negativamente el rendimiento de los modelos, ya que tienden a priorizar la clase mayoritaria.

El sobre-muestreo, como SMOTE, añade ejemplos sintéticos a la clase minoritaria para equilibrar la distribución de clases sin perder datos de la clase mayoritaria. Aunque esta técnica es útil, puede incrementar los tiempos de entrenamiento y el uso de memoria, además de aumentar el riesgo de sobreajuste, especialmente en conjuntos de datos pequeños. Por otro lado, el submuestreo, que reduce la clase mayoritaria, es más eficiente en cuanto a memoria y tiempo, pero puede llevar a la pérdida de información valiosa, por lo que no es recomendable en datasets pequeños.


-


Los modelos implementados fueron:

- **Regresión Logística (LR)**: Algoritmo supervisado para problemas de clasificación binaria que predice probabilidades según las características del conjunto de datos, aplicando una regresión lineal con una salida de 0 o 1.
  
- **k-Vecinos más Cercanos (kNN)**: Algoritmo utilizado para clasificación y regresión, que predice la variable de salida basada en los K vecinos más cercanos. El valor de K es clave para ajustar el modelo.

- **Árbol de Decisión (DTree)**: Algoritmo de clasificación y regresión que divide el conjunto de datos en subconjuntos basados en características, creando un árbol de decisiones que facilita las predicciones.

- **Bosque Aleatorio (RF)**: Algoritmo de ensamble que construye múltiples árboles de decisión en submuestras del conjunto de datos y promedia los resultados para mejorar la precisión y reducir el sobreajuste.

- **Extreme Gradient Boosting (XGBoost)**: Método basado en árboles de decisión que mejora el rendimiento mediante el refuerzo de gradientes. Cada árbol se construye para corregir los errores del anterior, continuando hasta que los residuales se minimicen o se alcance el número máximo de iteraciones.

- **Red Neuronal de Perceptrón Multicapa (MLP)**: Modelo inspirado en el funcionamiento del cerebro humano, compuesto por múltiples capas de neuronas conectadas que permiten el aprendizaje de patrones complejos.

- **Máquina de Vectores de Soporte (SVM)**: Algoritmo de clasificación que encuentra el hiperplano óptimo que separa las clases de datos de forma que maximiza la distancia entre los puntos más cercanos de ambas clases. También puede ser utilizado para regresión.


-
### **Análisis de Resultados:**

**Explicación de los valores de G-mean:**

La métrica **G-mean (Geometric Mean)** es una medida importante cuando se trabaja con datos desbalanceados, ya que equilibra la tasa de aciertos en las clases mayoritaria y minoritaria. G-mean es la raíz cuadrada del producto de las tasas de sensibilidad (recall) y especificidad. En otras palabras, penaliza modelos que predicen muy bien una clase (como la mayoritaria) a costa de predecir mal la otra (la minoritaria), ofreciendo un equilibrio entre ambas.

Valores cercanos a **1** indican un buen desempeño en ambas clases, mientras que valores más bajos indican que el modelo no está logrando un buen equilibrio entre la clase mayoritaria y la minoritaria.

- **Regresión Logística (LR)**:
	- G-mean Test: 0.816 (0.030)
	- G-mean Train: 0.832 (0.007)
  - Interpretación: La Regresión Logística muestra un buen equilibrio entre las clases en el conjunto de prueba (0.816) y entrenamiento (0.832), con una pequeña diferencia entre ambos. Esto sugiere que el modelo generaliza bien y no está sobreajustado, lo que es un resultado positivo.

- **k-Vecinos más Cercanos (kNN)**:
	- G-mean Test: 0.793 (0.021)
	- G-mean Train: 0.814 (0.006)
  - Interpretación: El kNN tiene un desempeño aceptable, aunque ligeramente inferior al de la Regresión Logística. La diferencia entre el G-mean de entrenamiento y prueba no es alarmante, pero sugiere que podría estar ligeramente sobreajustado, ya que rinde mejor en el conjunto de entrenamiento.

- **Árbol de Decisión (DTree)**:
	- G-mean Test: 0.791 (0.026)
	- G-mean Train: 0.887 (0.009)
  - Interpretación: El Árbol de Decisión muestra una diferencia considerable entre los resultados de prueba y entrenamiento, lo que sugiere sobreajuste. El modelo está aprendiendo demasiado bien el conjunto de entrenamiento (0.887) pero no generaliza tan bien al conjunto de prueba (0.791).

- **Bosque Aleatorio (RF)**:
	- G-mean Test: 0.813 (0.028)
	- G-mean Train: 0.970 (0.005)
  - Interpretación: El Bosque Aleatorio tiene un buen G-mean en el conjunto de prueba (0.813), pero un G-mean excesivamente alto en el conjunto de entrenamiento (0.970), lo que indica un claro sobreajuste. Este modelo ha aprendido demasiado bien los datos de entrenamiento y no generaliza tan bien.

- **XGBoost**:
	- G-mean Test: 0.816 (0.023)
	- G-mean Train: 0.970 (0.005)
  - Interpretación: XGBoost tiene uno de los mejores resultados en el conjunto de prueba (0.816), pero como el Bosque Aleatorio, su rendimiento en el conjunto de entrenamiento es extremadamente alto (0.970), lo que indica un sobreajuste significativo. A pesar de su buen desempeño, esto debe manejarse para evitar que el modelo se ajuste demasiado a los datos de entrenamiento.

- **Perceptrón Multicapa (MLP)**:
	- G-mean Test: 0.813 (0.026)
	- G-mean Train: 0.831 (0.008)
  - Interpretación: El MLP muestra un buen equilibrio con resultados similares en el conjunto de prueba (0.813) y entrenamiento (0.831). La diferencia entre ambos conjuntos es mínima, lo que sugiere que el modelo no está sobreajustado y generaliza bien.

- **Máquina de Vectores de Soporte (SVM)**:
	- G-mean Test: 0.822 (0.030)
	- G-mean Train: 0.840 (0.007)
  - Interpretación: La SVM tiene un desempeño sólido con un buen G-mean tanto en el conjunto de prueba (0.822) como en el de entrenamiento (0.840), lo que indica una buena capacidad de generalización y un equilibrio entre ambas clases.


-
### **Mejor Modelo:**

En conclusión, los modelos con mejor rendimiento y que demostraron ser los más robustos son:

- **Máquina de Vectores de Soporte (SVM)**: Este modelo ofrece un buen balance entre los resultados de prueba (0.822) y entrenamiento (0.840), lo que demuestra su capacidad para generalizar bien sin caer en el sobreajuste. Su habilidad para encontrar el hiperplano óptimo que maximiza la separación entre clases lo convierte en una opción confiable en contextos de clasificación con desbalance de clases.

- **Regresión Logística (LR-Elastic)**: La Regresión Logística también mostró un excelente desempeño, con un G-mean de 0.816 en el conjunto de prueba y 0.832 en el conjunto de entrenamiento. Este resultado indica que el modelo generaliza bien y no muestra señales de sobreajuste, lo que lo hace adecuado para problemas de clasificación como este.

Ambos modelos, SVM y LR, mostraron ser robustos, equilibrados y capaces de manejar adecuadamente el desbalance de clases. Son modelos confiables que pueden ser implementados en aplicaciones del mundo real, donde la precisión y la capacidad de generalización son cruciales.


-
### **Conclusión General:**

La actividad permitió comparar múltiples modelos y técnicas para abordar un problema de clasificación con datos desbalanceados en el contexto del riesgo crediticio. Aunque XGBoost mostró el mejor desempeño en términos de métricas en el conjunto de prueba, su tendencia al sobreajuste sugiere que no es la opción más confiable sin ajustes adicionales.

La Máquina de Vectores de Soporte (SVM) y la Regresión Logística emergen como los modelos más balanceados, ofreciendo un buen desempeño y capacidad de generalización. Esto resalta la importancia de considerar no solo las métricas de desempeño sino también la estabilidad y robustez del modelo al momento de seleccionar la mejor opción para implementación.

Finalmente, la experiencia destaca la relevancia de manejar adecuadamente el desbalance de clases y de utilizar técnicas de validación apropiadas para obtener modelos confiables y aplicables en contextos reales.


-
### **Referencias:**

- Alam, T. M., Shaukat, K., Hameed, I. A., Luo, S., Sarwar, M. U., Shabbir, S., Li, J., & Khushi, M. (2020). An investigation of credit card default prediction in the imbalanced datasets. IEEE Access, 8, 201173–201198. Disponible en: https://ieeexplore.ieee.org/document/9239944

- Anzai, Y. (2012). Pattern Recognition and Machine Learning. Morgan Kaufmann. Disponible en: https://biblioteca.tec.mx/oreilly

- Huyen, C. (s. f.). Designing machine learning systems. O’Reilly Online Learning. https://learning.oreilly.com/library/view/designing-machine-learning/9781098107956/ch04.html

- Abhishek, K., & Abdelaziz, M. (s. f.). Machine Learning for Imbalanced Data. O’Reilly Online Learning. https://learning.oreilly.com/library/view/machine-learning-for/9781801070836/B17259_01.xhtml#_idParaDest-20

- RandomForestClassifier. (s. f.). Scikit-learn. https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.RandomForestClassifier.html

- Cómo funciona el algoritmo XGBoost—ArcGIS Pro | Documentación. (s. f.). https://pro.arcgis.com/es/pro-app/latest/tool-reference/geoai/how-xgboost-works.htm#:~:text=XGBoost%20es%20la%20abreviatura%20de,aleatorio%20y%20refuerzo%20de%20gradientes.

- XGBoost Documentation — xgboost 2.1.1 documentation. (s. f.). https://xgboost.readthedocs.io/en/stable/

- Hu, Y., Yang, L., Chen, L., & Zhu, S. (2020). Research on a customer churn combination prediction model based on decision tree and neural network. En 2020 IEEE 5th International Conference on Cloud Computing and Big Data Analytics (ICCCBDA) (pp. 129–132). IEEE. https://doi.org/10.1109/ICCCBDA49378.2020.9095611

- Tecnológico de Monterrey. (2024). MNA_IAyAA_semana_5_y_6_Actividad_sept_2024.pdf. [Material de clase]. Curso Inteligencia Artificial y Aprendizaje Automático (Grupo 10), impartido por el Dr. Luis Eduardo Falcón Morales.

- Tecnológico de Monterrey. (2024). MNA_IAyAA_Titanic_semanas_5y6.ipynb. [Material complementario]. Curso Inteligencia Artificial y Aprendizaje Automático (Grupo 10), impartido por el Dr. Luis Eduardo Falcón Morales.

- Tecnológico de Monterrey. (2024). MNA_IAyAA_semana_6_Arboles_de_decision_y_BosqueAleatorio_teoria-1.pdf. [Material de clase]. Curso Inteligencia Artificial y Aprendizaje Automático (Grupo 10), impartido por el Dr. Luis Eduardo Falcón Morales.

- Universidad de California, Irvine. (s.f.). South German Credit Data Set. Recuperado en octubre de 2024 de https://archive.ics.uci.edu/dataset/522/south+german+credit


++++++++ Termina la sección de agregar texto +++++++++++

>> ### **Fin de la Actividad de las Semanas 5 y 6.**